### PR TITLE
Qwen Image

### PR DESCRIPTION
**Pull request by filipstrand**
Just scouting the terrain to begin with...

Initially this looks like a bigger task than I first expected (I mainly thought it would be a new text encoder, but it looks like there are a lot of other updates too).

The good news is that I was able to get the diffusers pipeline to work on my 32GB machine for a 128x128 image (but still quite slow to run), hopefully should be enough for debugging purposes... 

@anthonywu Once (if) the implementation is successful, it would be nice to discuss a little bit about how to approach the general idea of adding more non-flux models to this repo: I can see a couple of questions that are worth thinking about (should the repo be renamed - but then we loose the current "branding"?, How would general `mflux-` commands work etc?. How do we handle features for multiple models in a scalable way (e.g some things like img2img feels quite general and should be easy to support for multiple models, but other things, such as the current upscaling feature, is actually a Flux-specific ControlNet right now and would not work with Qwen etc).    

---

Here is an initial outline (fully autogenerated for now) of differences/similarities compared to the current MFLUX: 

# Qwen Image Port to MFlux - Implementation Guide

## Overview

This document outlines the comprehensive plan for porting Qwen Image (a text-to-image diffusion model) from the Diffusers PyTorch implementation to the MFlux MLX library. Qwen Image represents a significant architectural departure from Flux, requiring both component reuse and substantial new development.

### Key Differences: Flux vs Qwen Image

| Component | Flux | Qwen Image |
|-----------|------|------------|
| **Text Encoding** | Dual encoders (CLIP + T5) | Single VL model (Qwen2.5-VL) with prompt templating |
| **Transformer** | Standard MMDiT (joint → single blocks) | Dual-stream architecture (consistent joint processing) |
| **VAE Processing** | 2D spatial processing | 3D temporal processing with causal convolutions |
| **Position Embeddings** | Simple coordinate IDs (`EmbedND`) | Advanced 3D RoPE (`QwenEmbedRope`) |
| **Attention Pattern** | Joint attention then single attention | Dual-stream attention throughout all blocks |
| **Latent Handling** | Standard 16x compression | Special 2x2 patch packing/unpacking |

## 📁 Reference Implementation Structure

The reference implementation is available in `diffusers_impl/` with the following structure:

```
diffusers_impl/
├── shared/          # Reusable components (schedulers, attention, utilities)
├── flux/           # Flux reference implementation 
├── qwen/           # Qwen Image specific components
│   ├── pipeline_qwenimage.py      # Main pipeline
│   ├── transformer_qwenimage.py   # Dual-stream transformer
│   └── autoencoder_kl_qwenimage.py # 3D VAE with causal convolutions
└── README.md       # Architecture overview
```

### ✅ **REUSABLE Components** (40-50% of codebase)

**Low-Level Building Blocks:**
- `AttentionUtils` - Core attention computation logic
- Basic normalization patterns (RMSNorm, LayerNorm)
- Config system infrastructure
- Weight handling patterns
- Timestep embedding calculation logic
- General MLX tensor operations

**VAE Foundation Components:**
- `ResnetBlock2D` pattern → extend to `ResnetBlock3D`
- `UpSampler` logic → adapt for 3D temporal processing
- Basic convolution patterns → Conv2D to Conv3D adaptation
- Attention mechanism in VAE → reuse scaled dot product attention

### 🔧 **NEEDS PARAMETERIZATION** (30% of codebase)

**Currently Hardcoded for Flux:**
```python
# Examples of hardcoded values that need parameterization:
FeedForward: 3072 → 12288 → 3072
AdaLayerNormZero: 3072 input, 18432 output (6 chunks of 3072)
JointAttention: num_heads=24, head_dim=128, all 3072 dimensions
EmbedND: dim=3072, axes_dim=[16, 56, 56]
VAE Attention: 512 dimensions, 32 groups
```

**Parameterization Strategy:**
- Make dimensions configurable: `hidden_dim`, `ff_dim`, `num_heads`, `head_dim`
- Architecture parameters: `num_transformer_blocks`, block types
- RoPE parameters: `theta`, `axes_dim` for different models
- VAE parameters: channels, group sizes, temporal dimensions

### 🆕 **COMPLETELY NEW Components** (20% of codebase)

**Text Processing:**
- Qwen2.5-VL Text Encoder (major component)
- Prompt templating system with special tokens
- Attention mask handling for variable-length sequences

**Transformer Architecture:**
- `QwenImageTransformer` - Dual-stream architecture
- `QwenDoubleStreamAttnProcessor` - Specialized attention pattern
- 3D RoPE embeddings with temporal-spatial relationships
- Dual-stream transformer blocks

**VAE Components:**
- `QwenImageCausalConv3d` - Causal 3D convolutions with caching
- 3D ResNet blocks with temporal processing
- 3D Up/Down sampling with temporal dimension handling
- Latent packing/unpacking for 2x2 patches

### Phase 1: Core Infrastructure Setup
**Goal:** Basic project structure and configuration
- [ ] Create `QwenImageConfig` in `src/mflux/config/model_config.py`
- [ ] Add entry point `src/mflux/generate_qwen.py`
- [ ] Set up basic model configuration parameters
- [ ] Create placeholder classes for main components

### Phase 2: Component Parameterization  
**Goal:** Make existing Flux components reusable
- [ ] Parameterize `FeedForward` class (remove hardcoded dimensions)
- [ ] Parameterize `AdaLayerNormZero` and related normalization layers
- [ ] Make `JointAttention` configurable for different architectures
- [ ] Create configurable base classes for transformer blocks

### Phase 3: Transformer Architecture
**Goal:** Implement Qwen's dual-stream transformer
- [ ] Create `QwenImageTransformer` in `src/mflux/models/transformer/`
- [ ] Implement `QwenDoubleStreamAttnProcessor` 
- [ ] Port 3D RoPE embeddings (`QwenEmbedRope`)
- [ ] Create dual-stream transformer blocks
- [ ] Implement specialized attention pattern

### Phase 4: VAE Implementation
**Goal:** 3D temporal VAE with causal processing
- [ ] Create `QwenImageVAE` base class
- [ ] Implement `QwenImageCausalConv3d` with caching support
- [ ] Create 3D ResNet blocks with temporal processing
- [ ] Add 3D up/down sampling modules
- [ ] Implement latent packing/unpacking logic
- [ ] Add temporal consistency features

### Phase 5: Text Processing System
**Goal:** Qwen2.5-VL text encoder integration
- [ ] Port Qwen2.5-VL text encoder to MLX
- [ ] Implement prompt templating system
- [ ] Add attention mask handling
- [ ] Create `QwenPromptEncoder` following MFlux patterns
- [ ] Optimize memory usage for large VL model

### Phase 6: Integration & Pipeline
**Goal:** Complete generation pipeline
- [ ] Create main `QwenImage` class similar to `Flux1`
- [ ] Implement generation pipeline with dual-stream processing
- [ ] Add weight loading support for Qwen models
- [ ] Create `QwenImageInitializer` following `FluxInitializer` pattern
- [ ] Integrate all components into working pipeline

### Phase 7: Testing & Optimization
**Goal:** Validation and performance tuning
- [ ] Test with simple text-to-image prompts
- [ ] Validate against reference implementation outputs
- [ ] Memory optimization for 3D processing
- [ ] Performance benchmarking
- [ ] Documentation and examples

### 1. **PyTorch → MLX Conversion**
**Challenge:** 3D convolutions, complex attention mechanisms, RoPE embeddings
**Solution:** Leverage MLX's tensor operations, adapt convolution patterns, careful tensor manipulation

### 2. **Qwen2.5-VL Text Encoder**
**Challenge:** Large VL model requiring significant memory
**Solution:** Incremental porting, memory optimization, possible quantization support

### 3. **3D Temporal Processing**
**Challenge:** Causal convolutions with cache management
**Solution:** Implement efficient caching system, optimize memory usage

### 4. **Dual-Stream Architecture**
**Challenge:** Different from Flux's joint→single pattern
**Solution:** Careful attention mask handling, proper tensor routing

## 📊 Development Metrics

- **Estimated Reusability:** 40-50% of existing MFlux components
- **New Development:** 20% completely new components
- **Adaptation Required:** 30% existing components need parameterization
- **Critical Path:** Text encoder porting (most complex component)

## 🎯 Success Criteria

1. **Functional Pipeline:** Generate images from text prompts
2. **Output Validation:** Match reference implementation quality
3. **Performance:** Reasonable inference speed on Apple Silicon
4. **Memory Efficiency:** Handle typical image generation workloads
5. **Code Quality:** Follow MFlux architectural patterns and conventions

## 📚 Reference Materials

- **Diffusers Implementation:** `diffusers_impl/qwen/`
- **MFlux Patterns:** Existing Flux implementation in `src/mflux/flux/`
- **Architecture Guide:** `diffusers_impl/README.md`
- **Key Papers:** Qwen Image technical documentation

## 🚦 Getting Started

1. **Review Phase 1** tasks and set up basic infrastructure
2. **Start with parameterization** of existing components (Phase 2)
3. **Focus on transformer architecture** as the core differentiator (Phase 3)
4. **Tackle VAE implementation** once transformer is working (Phase 4)
5. **Integrate text encoder** as the final major component (Phase 5)

---

*This guide represents a comprehensive analysis of the Qwen Image porting effort. Adjust priorities and timeline based on development capacity and requirements.*

### CODE REVIEW

**Comment by mzbac** [comment]
Hi @filipstrand  there is a minor issue in the loading the weight but the generation works fine on my M2 Ultra around 54 Seconds to generate the 512x512 image..
image

image

qwen_output_42

**Comment by ivanfioravanti** [comment]
I will test this on M3 Ultra 512GB, @filipstrand would you like access to it to test and develop?

**Comment by shengkaixuan** [comment]
@mzbac  your pr works，amazing!
but the text rendering is not very good at the moment,
image

image

**Comment by mzbac** [comment]
Hi @filipstrand,I leave this PR to you. Feel free to copy the code from my fork when necessary, and do whatever you need to consolidate the stuff. You don't have to merge my PR, that was just for quick reference. I am currently working on porting the Qwen VAE encoder, which I think it is the only component left for supporting Qwen image edit. It's almost done, but it may take some time to refactor it to be consistent with your current VAE style. I will work on that in the background.

**Comment by filipstrand** [comment]
Some git problem, will reopen this again....
